stable_baselines3/her/replay_buffer.py

```diff
@@ -0,0 +1,323 @@
+        self.env = wrapped_env
+
+        self.observations = np.zeros((self.buffer_size, self.n_envs,) + self.obs_shape, dtype=np.float32)
+        if not self.add_her_while_sampling:
+            self.next_observations = np.zeros((self.buffer_size, self.n_envs,) + self.obs_shape, dtype=np.float32)
+
+        self.actions = np.zeros((self.buffer_size, self.n_envs, self.action_dim), dtype=np.float32)
+        self.rewards = np.zeros((self.buffer_size, self.n_envs), dtype=np.float32)
+        self.dones = np.zeros((self.buffer_size, self.n_envs), dtype=np.float32)
+        self.pos = 0
+        self.episode_transitions = []
+
+    def add(self, obs: np.ndarray,
+            next_obs: np.ndarray,
+            action: float,
+            reward: float,
+            done: bool):
+        """
+        add a new transition to the buffer
+
+        :param obs: (np.ndarray) the last observation
+        :param next_obs: (np.ndarray) the new observation
+        :param action: ([float]) the action
+        :param reward: (float) the reward of the transition
+        :param done: (bool) is the episode done
+        """
+        # Update current episode buffer
+        self.episode_transitions.append((obs, next_obs, action, reward, done))
+        if done:
+            # Add transitions to buffer only when an episode is over
+            self._store_episode()
+            # Reset episode buffer
+            self.episode_transitions = []
+
+    def _get_samples(self,
+                     batch_inds: np.ndarray,
+                     env: Optional[VecNormalize] = None
+                     ) -> ReplayBufferSamples:
+        if not self.add_her_while_sampling:
+            data = (self._normalize_obs(self.observations[batch_inds, 0, :], env),
+                    self.actions[batch_inds, 0, :],
+                    self._normalize_obs(self.next_observations[batch_inds, 0, :], env),
+                    self.dones[batch_inds],
+                    self._normalize_reward(self.rewards[batch_inds], env))
+            return ReplayBufferSamples(*tuple(map(self.to_torch, data)))
+        else:
+            '''
+            Sampling inspired by https://github.com/openai/baselines/blob/master/baselines/her/her_sampler.py
+            '''
+            batch_size = len(batch_inds)
+            her_inds = np.where(np.random.uniform(size=batch_size) < 1 - (1. / (1 + self.n_sampled_goal)))
+
+            obs_dict = self.env.convert_obs_to_dict(self.observations[batch_inds])
+            next_obs_dict = self.env.convert_obs_to_dict(self.observations[(batch_inds + 1) % self.buffer_size])
+
+            if self.goal_selection_strategy == GoalSelectionStrategy.RANDOM:
+                desired_goal_timestep_inds = np.random.randint(0, self.pos if not self.full else self.buffer_size,
+                                                               size=len(her_inds))
+            else:
+                end_inds = batch_inds.copy()
+                while True:
```

Comment:
The openai implementation indexes by episodes. I thought that might be inefficient when there are variable sized episodes. I'd tried it in an earlier commit ( e548af6dd74426decfaea2241edc5fad05a3ac19). Maybe that is closer to what you have in mind?

Reply:
yes, it is less efficient, unfortunately, there is no good way to avoid that (see #81 which is a bit of hack to re-use fixed size buffer to sample multi-step return)